File: apps/mobile-app/android/app/src/main/java/net/aliasvault/app/autofill/utils/RustItemMatcher.kt
package net.aliasvault.app.autofill.utils

import android.util.Log
import net.aliasvault.app.rustcore.JnaInitializer
import net.aliasvault.app.vaultstore.models.Item
import net.aliasvault.app.vaultstore.repositories.ItemWithCredentialInfo
import org.json.JSONArray
import org.json.JSONObject

/**
 * Wrapper for the Rust item matcher using UniFFI bindings.
 */
object RustItemMatcher {
    private const val TAG = "RustItemMatcher"

    /**
     * Matching mode for credential filtering.
     *
     * @property value The string value used in the Rust API.
     */
    enum class MatchingMode(val value: String) {
        /** Default matching mode - uses subdomain matching. */
        DEFAULT("default"),

        /** Exact URL matching only - no subdomain matching. */
        URL_EXACT("url_exact"),

        /** Subdomain matching - matches subdomains and root domains. */
        URL_SUBDOMAIN("url_subdomain"),
    }

    init {
        // Ensure JNA is initialized before any UniFFI calls
        JnaInitializer.ensureInitialized()
    }

    /**
     * Filter items based on app/website info using the Rust core credential matcher.
     *
     * @param items List of items to filter
     * @param searchText Search term (app package name, URL, or text)
     * @return Filtered list of items
     */
    fun filterItemsByAppInfo(
        items: List<Item>,
        searchText: String,
    ): List<Item> {
        // Early return for empty search
        if (searchText.isEmpty()) {
            return items
        }

        try {
            // Convert items to JSON format expected by Rust
            val rustCredentials = JSONArray()
            val itemMap = mutableMapOf<String, Item>()

            for (item in items) {
                val idString = item.id.toString()
                val urlsArray = JSONArray()
                item.urls.forEach { urlsArray.put(it) }
                val credJson = JSONObject().apply {
                    put("Id", idString)
                    put("ItemName", item.name ?: JSONObject.NULL)
                    put("ItemUrls", urlsArray)
                    put("Username", item.username ?: JSONObject.NULL)
                }
                rustCredentials.put(credJson)
                itemMap[idString] = item
            }

            // Prepare input JSON for Rust
            val input = JSONObject().apply {
                put("credentials", rustCredentials)
                put("current_url", searchText)
                put("page_title", "")
                put("matching_mode", "default")
            }

            // Call Rust via UniFFI
            val outputJson = uniffi.aliasvault_core.filterCredentialsJson(input.toString())

            // Parse output
            val output = JSONObject(outputJson)
            val matchedIds = output.getJSONArray("matched_ids")

            // If no matches found, return empty list
            if (matchedIds.length() == 0) {
                return emptyList()
            }

            // Convert matched IDs back to items, maintaining order
            val filtered = mutableListOf<Item>()
            for (i in 0 until matchedIds.length()) {
                val id = matchedIds.getString(i)
                itemMap[id]?.let { filtered.add(it) }
            }

            return filtered
        } catch (e: Exception) {
            Log.e(TAG, "Error filtering items with Rust matcher: ${e.message}", e)
            // Fallback to returning all items on error
            return items
        }
    }

    /**
     * Filter ItemWithCredentialInfo items based on rpId using the Rust core credential matcher.
     * Used during passkey registration to find existing credentials that can have a passkey merged into them.
     *
     * @param items List of items to filter (without passkeys).
     * @param rpId The relying party identifier (domain) to match against.
     * @param rpName The relying party name (used for title matching fallback).
     * @param matchingMode The matching mode to use (default: URL_SUBDOMAIN).
     * @return List of items that match the rpId, in priority order.
     */
    fun filterItemsForPasskeyMerge(
        items: List<ItemWithCredentialInfo>,
        rpId: String,
        rpName: String? = null,
        matchingMode: MatchingMode = MatchingMode.URL_SUBDOMAIN,
    ): List<ItemWithCredentialInfo> {
        // Early return for empty rpId or items
        if (rpId.isEmpty() || items.isEmpty()) {
            return emptyList()
        }

        try {
            // Convert items to JSON format expected by Rust
            val rustCredentials = JSONArray()
            val itemMap = mutableMapOf<String, ItemWithCredentialInfo>()

            for (item in items) {
                val idString = item.itemId.toString()
                val urlsArray = JSONArray()
                item.urls.forEach { urlsArray.put(it) }

                val credJson = JSONObject().apply {
                    put("Id", idString)
                    put("ItemName", item.serviceName ?: JSONObject.NULL)
                    put("ItemUrls", urlsArray)
                    put("Username", item.username ?: JSONObject.NULL)
                }
                rustCredentials.put(credJson)
                itemMap[idString] = item
            }

            // Prepare input JSON for Rust
            // Use https:// prefix for the rpId to match URL format
            val input = JSONObject().apply {
                put("credentials", rustCredentials)
                put("current_url", "https://$rpId")
                put("page_title", rpName ?: "")
                put("matching_mode", matchingMode.value)
            }

            // Call Rust via UniFFI
            val outputJson = uniffi.aliasvault_core.filterCredentialsJson(input.toString())

            // Parse output
            val output = JSONObject(outputJson)
            val matchedIds = output.getJSONArray("matched_ids")

            // If no matches found, return empty list
            if (matchedIds.length() == 0) {
                return emptyList()
            }

            // Convert matched IDs back to items, maintaining Rust's priority order
            val result = mutableListOf<ItemWithCredentialInfo>()
            for (i in 0 until matchedIds.length()) {
                val id = matchedIds.getString(i)
                itemMap[id]?.let {
                    result.add(it)
                }
            }

            return result
        } catch (e: Exception) {
            Log.e(TAG, "Error filtering items for passkey merge: ${e.message}", e)
            // Return empty list on error
            return emptyList()
        }
    }
}

File: apps/mobile-app/ios/VaultStoreKit/Utils/RustItemMatcher.swift
import Foundation
import RustCoreFramework

/// Wrapper for the Rust item matcher using UniFFI bindings.
/// Used in VaultStoreKit for passkey registration to find mergeable items.
public enum RustItemMatcher {

    /// Matching mode for item filtering.
    public enum MatchingMode: String {
        case `default` = "default"
        case urlExact = "url_exact"
        case urlSubdomain = "url_subdomain"
    }

    /// Filter items based on rpId using the Rust core item matcher.
    /// Used during passkey registration to find existing items that can have a passkey merged into them.
    ///
    /// - Parameters:
    ///   - items: List of items to filter (without passkeys)
    ///   - rpId: The relying party identifier (domain) to match against
    ///   - matchingMode: The matching mode to use (default: urlSubdomain)
    /// - Returns: List of item IDs that match the rpId
    public static func filterItems(
        _ items: [ItemWithCredentialInfoData],
        rpId: String,
        matchingMode: MatchingMode = .urlSubdomain
    ) -> [UUID] {
        // Early return for empty rpId or items
        if rpId.isEmpty || items.isEmpty {
            return []
        }

        do {
            // Convert items to the format expected by Rust
            let rustCredentials = items.map { item -> [String: Any?] in
                return [
                    "Id": item.itemId.uuidString,
                    "ItemName": item.serviceName as Any?,
                    "ItemUrls": item.urls,
                    "Username": item.username as Any?
                ]
            }

            // Prepare input JSON for Rust
            // Use https:// prefix for the rpId to match URL format
            let input: [String: Any] = [
                "credentials": rustCredentials,
                "current_url": "https://\(rpId)",
                "page_title": "",
                "matching_mode": matchingMode.rawValue
            ]

            let inputData = try JSONSerialization.data(withJSONObject: input, options: [])
            guard let inputJson = String(data: inputData, encoding: .utf8) else {
                print("[RustItemMatcher] Failed to create input JSON")
                return []
            }

            // Call Rust via UniFFI
            let outputJson = try filterCredentialsJson(inputJson: inputJson)

            // Parse output
            guard let outputData = outputJson.data(using: .utf8),
                  let output = try JSONSerialization.jsonObject(with: outputData) as? [String: Any],
                  let matchedIds = output["matched_ids"] as? [String] else {
                print("[RustItemMatcher] Failed to parse output JSON")
                return []
            }

            // If no matches found, return empty array
            if matchedIds.isEmpty {
                return []
            }

            // Convert matched IDs back to UUIDs, maintaining order
            return matchedIds.compactMap { UUID(uuidString: $0) }

        } catch {
            print("[RustItemMatcher] Error filtering items: \(error)")
            // Return empty array on error
            return []
        }
    }

    /// Filter items and return the full ItemWithCredentialInfoData objects.
    /// Convenience method that returns filtered items instead of just IDs.
    ///
    /// - Parameters:
    ///   - items: List of items to filter (without passkeys)
    ///   - rpId: The relying party identifier (domain) to match against
    ///   - matchingMode: The matching mode to use (default: urlSubdomain)
    /// - Returns: List of items that match the rpId, in priority order
    public static func filterItemsWithData(
        _ items: [ItemWithCredentialInfoData],
        rpId: String,
        matchingMode: MatchingMode = .urlSubdomain
    ) -> [ItemWithCredentialInfoData] {
        let matchedIds = filterItems(items, rpId: rpId, matchingMode: matchingMode)

        // Map matched IDs back to items, maintaining Rust's priority order
        return matchedIds.compactMap { matchedId in
            items.first { $0.itemId == matchedId }
        }
    }
}
